Zend/zend_alloc.c

```diff
@@ -3009,6 +3091,206 @@ ZEND_API void zend_mm_get_custom_handlers(zend_mm_heap *heap,
+	heap->use_custom_heap |= ZEND_MM_CUSTOM_HEAP_OBSERVED;
+#endif
+}
+
+/*
+ * This function will shutdown the observers for the current heap and free
+ * global structures. This will be called before PHP module shutdown phase.
+ */
+void zend_mm_observers_shutdown(zend_mm_heap *heap)
+{
+#if ZEND_MM_CUSTOM
+	if (heap == NULL) {
+		heap = AG(mm_heap);
+		if (heap == NULL) {
+			return;
+		}
+	}
+	zend_mm_heap_observer *current_heap_observer = heap->observers;
+	zend_mm_heap_observer *next_heap_observer = NULL;
+	while (current_heap_observer != NULL) {
+		next_heap_observer = current_heap_observer->next;
+		pefree(current_heap_observer, 1);
+		current_heap_observer = next_heap_observer;
+	}
+	heap->observers = NULL;
+	return;
+#endif
+}
+
+void zend_mm_observers_unregister(void)
+{
+#if ZEND_MM_CUSTOM
+	zend_mm_observer *current = zend_mm_observers;
+	zend_mm_observer *next = NULL;
+	while (current != NULL) {
+		next = current->next;
+		pefree(current, 1);
+		current = next;
+	}
+	zend_mm_observers = NULL;
+#endif
+}
+
+/*
+ * This internal function will check if there are still enabled observer on the
+ * heap. In case there are, it will set the `ZEND_MM_CUSTOM_HEAP_OBSERVED` flag
+ * otherwise it will unset the flag.
+ */
+void zend_mm_observer_sync_flag(bool enabled_observer) {
+#if ZEND_MM_CUSTOM
+	zend_mm_heap *heap = AG(mm_heap);
+	if (heap == NULL) {
+		return;
+	}
+	if (enabled_observer) {
+		// When an observer was enabled, we don't need to find if there were
+		// enabled observers
+		heap->use_custom_heap |= ZEND_MM_CUSTOM_HEAP_OBSERVED;
+		return;
+	}
+	// An observer got disabled, so we need to check if there are still enabled
+	// observers left in the list
+	zend_mm_heap_observer *current = heap->observers;
+	while (current != NULL) {
+		if (current->enabled == true) {
+			// in case we find an enabled observer there is nothing to do
+			return;
+		}
+		current = current->next;
+	}
+	heap->use_custom_heap &= ~ZEND_MM_CUSTOM_HEAP_OBSERVED;
+#endif
+}
+
+ZEND_API bool zend_mm_observer_enabled(zend_mm_heap *heap, zend_mm_observer *node) {
+#if ZEND_MM_CUSTOM
+	if (heap == NULL) {
+		heap = AG(mm_heap);
+		if (heap == NULL) {
+			return false;
+		}
+	}
+	zend_mm_heap_observer *current = heap->observers;
+	while (current != NULL) {
+		if (current->observer == node) {
+			return current->enabled;
+		}
+		current = current->next;
+	}
+#endif
+	return false;
+}
+
+ZEND_API bool zend_mm_observer_enable(zend_mm_heap *heap, zend_mm_observer *node) {
+#if ZEND_MM_CUSTOM
+	if (heap == NULL) {
+		heap = AG(mm_heap);
+		if (heap == NULL) {
+			return false;
+		}
+	}
+	zend_mm_heap_observer *current = heap->observers;
+	while (current != NULL) {
+		if (current->observer == node) {
+			current->enabled = true;
+			zend_mm_observer_sync_flag(true);
+			return true;
+		}
+		current = current->next;
+	}
+#endif
+	return false;
+}
+
+ZEND_API bool zend_mm_observer_disable(zend_mm_heap *heap, zend_mm_observer *node) {
```

Comment:
This could be unified with `zend_mm_observer_enable` as the code is essentially the same. You can add convenience functions that pass `true` of `false` if you prefer that API.